tsl/src/chunk.c

```diff
@@ -561,3 +562,206 @@ chunk_unfreeze_chunk(PG_FUNCTION_ARGS)
 	bool ret = ts_chunk_unset_frozen(chunk);
 	PG_RETURN_BOOL(ret);
 }
+
+static List *
+chunk_id_list_create(ArrayType *array)
+{
+	/* create a sorted list of chunk ids from array */
+	ArrayIterator it;
+	Datum id_datum;
+	List *id_list = NIL;
+	bool isnull;
+
+	it = array_create_iterator(array, 0, NULL);
+	while (array_iterate(it, &id_datum, &isnull))
+	{
+		if (isnull)
+			ereport(ERROR,
+					(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+					 errmsg("chunks array arguments cannot be NULL")));
+		id_list = lappend_int(id_list, DatumGetInt32(id_datum));
+	}
+	array_free_iterator(it);
+
+	(void) list_sort_compat(id_list, list_int_cmp_compat);
+	return id_list;
+}
+
+static List *
+chunk_id_list_exclusive_right_merge_join(const List *an_list, const List *dn_list)
+{
+	/*
+	 * merge join two sorted list and return only values which exclusively
+	 * exists in the right target (dn_list list)
+	 */
+	List *result = NIL;
+	const ListCell *l = list_head(an_list);
+	const ListCell *r = list_head(dn_list);
+	for (;;)
+	{
+		if (l && r)
+		{
+			int compare = list_int_cmp_compat(l, r);
+			if (compare == 0)
+			{
+				/* l = r */
+				l = lnext_compat(an_list, l);
+				r = lnext_compat(dn_list, r);
+			}
+			else if (compare < 0)
+			{
+				/* l < r */
+				/* chunk exists only on the access node */
```

Comment:
Should this case every happen? Seems odd that it would. Should we report error here (corrupted metadata)?